docs: add the TAD (#74)
thomas-mauran/LinkedOut: README.md

# LinkedOut

## Introduction

LinkedOut is a user-friendly platform tailored to seasonal workers on the hunt for job opportunities.

This platform offers a variety of essential features, including user profile management, job listing services, intelligent job recommendations, streamlined recruitment processes, transparent review and rating systems, and multilingual support (initially in French and English).

## Microservices

One of the main goal of this project was to learn how to create and manage a microservice architecture.

> ![microservices](./docs/ms.png)

Reviewer: ```suggestion
![Architecture of the platform](./docs/ms.png)
```

Within our microservices architecture, we have implemented a centralized authentication mechanism using Keycloak and the API Gateway. The API Gateway intercepts incoming requests from clients and checks the JWT provided in the request headers against the Keycloak server to ensure the user's authentication and authorization status. If the JWT is valid, the API Gateway forwards the request to the appropriate microservice for further processing.
We decided to only check the token in the gateway and not in each microservice to avoid having to check the token in each microservice. We could have do a zero trust architecture where each microservice would check the token but we decided to go with the centralized approach.

We use a custom keycloak config stored in [linked_realm](./config/keycloak/linkedout_realm.json) which contains the roles and the clients we use. This config is then imported in the keycloak server and saves us a lot of time since we don't have to create the roles and the clients manually.

### Profile Service

> The Profile Service is responsible for managing user profiles within the LinkedOut platform. It handles the creation, retrieval, and modification of user profiles, as well as the management of user experiences, evaluation, and availability. This service also has to contact the recommendation service whenever a new user is created to store a neo4j node with his uuid to then later recommend special job offers for this user.

Reviewer: ```suggestion
The Profile Service is responsible for managing user profiles within the LinkedOut platform. It handles the creation, retrieval, and modification of user profiles, as well as the management of user experiences, evaluation, and availability. 

This service also has to contact the recommendation service whenever a new user is created to store a Neo4j node with their UUID to then later recommend special job offers for this user.
```

### Job Service

The Job Service is responsible for managing job offers within the LinkedOut platform. It does not handle the creating of jobs and jobOffers since it was specified the "employer service" was responsible for that. It handles the retrieval of job offers, jobs and jobCategories, the creation of job applications, and the management of job applications and job offers.
This service also has to contact the rsecommendation service to retrieve custom job offers for a user. We used flyway migration to keep the database up to date and store the constant informations like job categories and jobs as well as some jobOffers since we don't have any routes to create those.

### Message Service

- Explain the role of the message services

### Notification Service

- Explain the role of the notification services

### Recommendation Service

The Recommendation Service is responsible for providing personalized job recommendations to users based on their profiles and experiences. It leverages a graph database (Neo4j) to store and manage user profiles, experiences, and job offers, and uses graph-based algorithms to generate recommendations.

> Like in the job service we populate the database with the job categories, job offers and jobs. To do so we use a [cypher script](./backend/recommendation/) and a neo4j plugin named APOC (Awesome Procedures on Cypher), which provides a set of procedures and functions to enhance the capabilities of Neo4j.

Reviewer: ```suggestion
Like in the job service, we populate the database with the job categories, job offers and jobs. To do so, we use a [Cypher script](./backend/recommendation/) and a Neo4j plugin named APOC (Awesome Procedures On Cypher), which provides a set of procedures and functions to enhance the capabilities of Neo4j.
```

> ![graph db picture](./docs/assets/graph-db-1.png)

Reviewer: ```suggestion
![Graph database visualization](./docs/assets/graph-db-1.png)
```

> In the picture above you can see the graph database after initialization. In red you see each job nodes, in yellow each job category and in blue each job offer nodes.

Reviewer: ```suggestion
In the picture above, you can see the graph database after initialization: in red there's each job node, in yellow each job category node and in blue each job offer node.
```

> When a new user profile is created, the Recommendation Service stores the user's profile as a new node in the Neo4j graph database containing only it's id. Same thing when a new experience is created. We create a new node with the experience id and link it to both a user node and a job. By doing so we have a graph of users, experiences, jobs and job offers. We then use the neo4j driver to retrieve the recommendations. We use the graph database to find the shortest path between the user and the job offer. We then return a list of job offers ids. This list of job offers ids is then used to retrieve the job offers from the job service. If there is no job offers to return the job service will return all the job offers in the database.
> ![graph db picture](./docs/assets/graph-db-2.png)

Reviewer: ```suggestion
When a new user profile is created, the Recommendation Service stores the user profile as a new node in the Neo4j graph database containing only its ID. Same thing when a new experience is created: we create a new node with the experience ID and link it to both a user node and a job. 

By doing so we have a graph of users, experiences, jobs and job offers. We then use the Neo4j driver to retrieve the recommendations. We use the graph database to find the shortest path between the user and the job offer. We then return a list of job offers IDs. This list of job offers IDs is then used to retrieve the job offers from the job service. If there is no job offers to return, the job service will return all the job offers in the database.

![Graph database visualization](./docs/assets/graph-db-2.png)
```
